renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp

```diff
@@ -4546,6 +4546,53 @@ void WrappedVulkan::vkCmdCopyQueryPoolResults(VkCommandBuffer commandBuffer, VkQ
     }
     record->MarkBufferFrameReferenced(GetRecord(destBuffer), destOffset, size,
                                       eFrameRef_PartialWrite);
+
+    const QueryPoolInfo *qpInfo = GetRecord(queryPool)->queryPoolInfo;
+    if(qpInfo)
+    {
+      VkMemoryBarrier barrier = {
+          VK_STRUCTURE_TYPE_MEMORY_BARRIER,
+          NULL,
+          VK_ACCESS_TRANSFER_WRITE_BIT,
+          VK_ACCESS_TRANSFER_READ_BIT,
+      };
+      vkCmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_TRANSFER_BIT,
+                           VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, 0, 1, &barrier, 0, VK_NULL_HANDLE, 0,
+                           VK_NULL_HANDLE);
+
+      const VkDeviceSize resultSize =
+          (flags & VK_QUERY_RESULT_64_BIT) ? sizeof(uint64_t) : sizeof(uint32_t);
+
+      // If the stride matches the source buffer, then we can copy everything in a single region
+      if((flags & VK_QUERY_RESULT_64_BIT) && destStride == resultSize)
+      {
+        const VkBufferCopy region = {
+            firstQuery * resultSize,
+            destOffset,
+            queryCount * resultSize,
+        };
+        vkCmdCopyBuffer(commandBuffer, qpInfo->m_Buffer.buf, destBuffer, 1, &region);
+      }
+      else
+      {
+        // Copy each result into the destination, converting if required
+        rdcarray<VkBufferCopy> regions;
+        regions.reserve(queryCount);
+        for(size_t i = 0; i < queryCount; ++i)
+          regions.push_back(VkBufferCopy{(firstQuery + i) * sizeof(uint64_t),
+                                         destOffset + (i * destStride), resultSize});
+
+        vkCmdCopyBuffer(commandBuffer, qpInfo->m_Buffer.buf, destBuffer, (uint32_t)regions.size(),
+                        regions.data());
+      }
+
+      if(flags & VK_QUERY_RESULT_WITH_AVAILABILITY_BIT)
```

Comment:
> Again though I would not expect anyone to actually use this

Yeah but the spec says they can and if someone does they'll read past the end of the results array (because it's half the size they expect) and complain.

Reply:
Definitely, I just mean it's not a problem to put it in the suboptimal bucket with non-64bit results by doing separate copies/updates.